packages/teams-js/src/internal/performanceStatistics.ts

```diff
@@ -0,0 +1,114 @@
+      throw errorNotSupportedOnPlatform;
+    }
+    this.distributionBins = new Uint32Array(distributionBinSize);
+  }
+
+  /**
+   * Call this function before processing every frame
+   */
+  public processStarts(effectId: string, frameWidth: number, frameHeight: number): void {
+    if (!this.suitableForThisSession(effectId, frameWidth, frameHeight)) {
+      this.reportAndResetSession(this.getStatistics(), effectId, frameWidth, frameHeight);
+    }
+    this.start();
+  }
+
+  public processEnds(): void {
+    // calculate duration of the process and record it
+    const durationInMillisecond = performance.now() - this.frameProcessingStartedAt;
+    const binIndex = Math.floor(Math.max(0, Math.min(this.distributionBins.length - 1, durationInMillisecond)));
+    this.distributionBins[binIndex] += 1;
+    this.sampleCount += 1;
+  }
+
+  private getStatistics(): PerformanceStatisticsResult {
+    if (!this.currentSession) {
+      return null;
+    }
+    return {
+      effectId: this.currentSession.effectId,
+      frameHeight: this.currentSession.frameHeight,
+      frameWidth: this.currentSession.frameWidth,
+      duration: performance.now() - this.currentSession.startedAtInMs,
+      sampleCount: this.sampleCount,
+      distributionBins: this.distributionBins.slice(),
+    };
+  }
+
+  private start(): void {
+    this.frameProcessingStartedAt = performance.now();
+  }
+
+  private suitableForThisSession(effectId: string, frameWidth: number, frameHeight: number): boolean {
+    return (
+      this.currentSession &&
+      this.currentSession.effectId === effectId &&
+      this.currentSession.frameWidth === frameWidth &&
+      this.currentSession.frameHeight === frameHeight
+    );
+  }
+
+  private reportAndResetSession(result, effectId, frameWidth, frameHeight): void {
+    result && this.report(result);
+    this.resetCurrentSession(this.getNextTimeout(effectId, this.currentSession), effectId, frameWidth, frameHeight);
+    if (this.timeoutId) {
+      window.clearTimeout(this.timeoutId);
+    }
+    this.timeoutId = window.setTimeout(
+      (() => this.reportAndResetSession(this.getStatistics(), effectId, frameWidth, frameHeight)).bind(this),
+      this.currentSession.timeoutInMs,
+    );
+  }
+
+  private resetCurrentSession(timeoutInMs: number, effectId: string, frameWidth: number, frameHeight: number): void {
+    this.currentSession = {
+      startedAtInMs: performance.now(),
+      timeoutInMs,
+      effectId,
+      frameWidth,
+      frameHeight,
+    };
+    this.sampleCount = 0;
+    this.distributionBins.fill(0);
+  }
+
+  // send the statistics result every n second, where n starts from 1, 2, 4...and finally stays at every 30 seconds.
+  private getNextTimeout(effectId: string, currentSession?: { timeoutInMs: number; effectId: string }): number {
+    // only reset timeout when new session or effect changed
+    if (!currentSession || currentSession.effectId !== effectId) {
+      return 1000;
```

Comment:
nit: This 1000 seems like it should be a named const. Seems possible you'd want to quickly adjust how long the timeout is in the future. Possibly same for 1000*30 below?